### PR TITLE
Implement path finalization for external manifest store

### PR DESCRIPTION
**Pull request by chebbyChefNEQ**
In #1207 we made an update to the commit procedure. At the end of the commit process we want to repoint the manifest path to the finalized path `{x}.manifest` from `{x}.manifest-{uuid}`. This PR implements this new step.

A few changes that happened because of the update
* we used to check the object store for `{x}.manifest` as the sign that the commit has finalized. Now the code check the external store and if the returns path is does not end in UUID, we assume the manifest is finalized and does not require syncing.
* `ExternalManifestStore::get_latest_version` now returns both the version number and the path to the manifest
* Fix a few more non-atomic copys by using copy-rename. (Will undo these once we upgrade `object_store` to a version with atomic `copy` operation)

### CODE REVIEW

File: rust/src/io/commit.rs [hunk 1]
```diff
@@ -88,20 +96,22 @@ async fn list_manifests<'a>(
...
-    commit_handler: &dyn CommitHandler,
     from_path: &Path,
     base_path: &Path,
     object_store: &ObjectStore,
 ) -> Result<()> {
-    let latest_path = commit_handler
-        .resolve_latest_version(base_path, object_store)
-        .await?;
+    let latest_path = latest_manifest_path(base_path);
     // Then, create a copy at the latest manifest path
```
Comment (wjones127) [inline rust/src/io/commit.rs:104]: nit: this comment feels a little about of place in the sequence. Should it be moved down a few lines?

File: rust/src/io/commit/external_manifest.rs [hunk 2]
```diff
@@ -46,11 +49,14 @@ pub trait ExternalManifestStore: std::fmt::Debug + Send + Sync {
     /// Get the manifest path for a given base_uri and version
     async fn get(&self, base_uri: &str, version: u64) -> Result<String>;
 
-    /// Get the latest version of a dataset at the base_uri
-    async fn get_latest_version(&self, base_uri: &str) -> Result<Option<u64>>;
+    /// Get the latest version of a dataset at the base_uri, and the path to the manifest
```
Comment (wjones127) [inline rust/src/io/commit/external_manifest.rs:52]: Why do we need the path? Shouldn't it be uniquely determined by the version? Or is this just an optimization?

**Review by wjones127 — CHANGES_REQUESTED** [review]
One question

File: rust/src/io/commit/external_manifest.rs (hunk 2 above)
Comment (chebbyChefNEQ) [reply rust/src/io/commit/external_manifest.rs:52]: This is an optimization to not have to `get(get_latest_version)`

Comment (wjones127) [reply rust/src/io/commit/external_manifest.rs:52]: Got it. Let's specify that in the docs:
```suggestion
    /// Get the latest version of a dataset at the base_uri, and the path to the manifest.
    /// The path is provided as an optimization. The path is deterministic based on
    /// the version and the store should not customize it.
```

Comment (chebbyChefNEQ) [reply rust/src/io/commit/external_manifest.rs]: done

**Review by wjones127 — APPROVED** [review]

**Pull request merged**